### PR TITLE
 pkg/apis/monitoring/v1: Add validations for log settings at CRD level

### LINKED ISSUE #4524: Improve documentation and move possible validations via OpenAPI
**What is missing?**

While working on adding [validation for scrapeTimeout](https://github.com/prometheus-operator/prometheus-operator/pull/4491#discussion_r782103988), we noted that its a bit confusing and documentation is not clear how default values are set for CRD fields. We need to improve documentation on how default values are set and move validations possible using [OpenAPI](https://book.kubebuilder.io/reference/markers/crd-validation.html)

**Why do we need it?**

This is needed for better understanding and troubleshooting incase of issues

**Environment**

NA

* Prometheus Operator version:

    `Insert image tag or Git SHA here`
    
    

0.53.1

**Anything else we need to know?**:

See discussion in the PR #4491

### PR DESCRIPTION
**Pull request by slashpai**
pkg/apis/monitoring/v1: Add validations for log settings at CRD level

Related-to #4524

Signed-off-by: Jayapriya Pai <slashpai9@gmail.com>

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

pkg/apis/monitoring/v1: Add validations for log settings at CRD level

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
pkg/apis/monitoring/v1: Add validations for log settings at CRD level
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -125,14 +125,19 @@ type CommonPrometheusFields struct {
 	// _not_ be added when value is set to empty string (`""`).
 	PrometheusExternalLabelName *string `json:"prometheusExternalLabelName,omitempty"`
 	// Log level for Prometheus to be configured with.
+	//+kubebuilder:validation:Enum=debug;info;warn;error
 	LogLevel string `json:"logLevel,omitempty"`
 	// Log format for Prometheus to be configured with.
+	//+kubebuilder:validation:Enum=logfmt;json
 	LogFormat string `json:"logFormat,omitempty"`
 	// Interval between consecutive scrapes. Default: `1m`
+	// +kubebuilder:validation:Pattern:="^[0-9]+(ms|s|m|h|d|w|y)$"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:134]: model.Duration allows a value of 0 (from [here](https://github.com/prometheus/common/blob/0e1254b764d0c37ae98f7b4b130b22c42844c92d/model/time.go#L192-L194)) as well as more complex durations (e.g. `1m2s`).

```suggestion
	// +kubebuilder:validation:Pattern:="^0|(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?$"
	// +kubebuilder:default:=30s
```

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -125,14 +125,19 @@ type CommonPrometheusFields struct {
 	// _not_ be added when value is set to empty string (`""`).
 	PrometheusExternalLabelName *string `json:"prometheusExternalLabelName,omitempty"`
 	// Log level for Prometheus to be configured with.
+	//+kubebuilder:validation:Enum=debug;info;warn;error
 	LogLevel string `json:"logLevel,omitempty"`
 	// Log format for Prometheus to be configured with.
+	//+kubebuilder:validation:Enum=logfmt;json
 	LogFormat string `json:"logFormat,omitempty"`
 	// Interval between consecutive scrapes. Default: `1m`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:133]: the default applied by the operator is 30s if not set.

File: pkg/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -699,8 +705,10 @@ type ThanosSpec struct {
 	// Maps to the '--grpc-server-tls-*' CLI args.
 	GRPCServerTLSConfig *TLSConfig `json:"grpcServerTlsConfig,omitempty"`
 	// LogLevel for Thanos sidecar to be configured with.
+	//+kubebuilder:validation:Enum=debug;info;warn;error
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:708]: the empty value is also valid

File: pkg/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -699,8 +705,10 @@ type ThanosSpec struct {
...
 	LogLevel string `json:"logLevel,omitempty"`
 	// LogFormat for Thanos sidecar to be configured with.
+	//+kubebuilder:validation:Enum=logfmt;json
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:711]: same here

**Review by simonpasquier — COMMENTED** [review]
I did a quick pass but maybe you can split the work in several PRs and group by field types (e.g. log settings, size settings, time intervals, ...)

**Comment by slashpai** [comment]
For all alertmanager time settings since it use [kingpin](https://github.com/prometheus/alertmanager/blob/d155153305afc25471b9879928d8d93df77b12a8/cmd/alertmanager/main.go#L209-L217) for cli flags which in-turn uses [time.Duration](https://github.com/alecthomas/kingpin/blob/17f40c25f4977ef446a79e34e46270665dffa263/parsers.go#L36-L41) which allows [till hours](https://pkg.go.dev/time#Duration) only

**Comment by slashpai** [comment]
@simonpasquier I updated with changes. I have tried to cover fields in pkg/apis/monitoring/v1/types.go as part of this PR

**Comment by simonpasquier** [comment]
i think you need to remove 1ea174e4a15d0ee366ef43e6f2f387856a83e8d7 from this pull request :)

**Comment by slashpai** [comment]
> i think you need to remove [1ea174e](https://github.com/prometheus-operator/prometheus-operator/commit/1ea174e4a15d0ee366ef43e6f2f387856a83e8d7) from this pull request :)

ya I had rebased branch missed to see that here. I will update :)

**Comment by slashpai** [comment]
Rebased branch again

**Review by simonpasquier — APPROVED** [review]
lgtm

**Pull request merged**